src/contrib/mlir/pass/mlir_subgraph_extraction.cpp

```diff
@@ -28,52 +29,148 @@ using namespace ngraph::pass;
 
 #define TI(x) std::type_index(typeid(x))
 
+void MLIRSubgraphExtractionPass::MLIRSubgraph::add_inputs(NodeVector &inputs)
+{
+    // inputs list are not exclusive, avoid duplication
+    for (NodeVector::iterator it = inputs.begin(); it != inputs.end(); it++)
+    {
+        std::shared_ptr<Node> node = *it;
+        if (std::find(m_input_nodes.begin(), m_input_nodes.end(), node) == m_input_nodes.end())
+        {
+            m_input_nodes.push_back(node);
+        }
+    }
+}
+
+void MLIRSubgraphExtractionPass::MLIRSubgraph::add_node(std::shared_ptr<Node> node)
+{
+    NGRAPH_CHECK(std::find(m_nodes.begin(), m_nodes.end(), node) == m_nodes.end(), "node added to graph before");
+    m_nodes.push_back(node);
+}
+
+void MLIRSubgraphExtractionPass::MLIRSubgraph::merge(MLIRSubgraph& other)
+{
+    // nodes  of sub-graphs are exclusive
+    m_nodes.insert(m_nodes.end(), other.get_nodes().begin(), other.get_nodes().end());
+    // merge inputs
+    add_inputs(other.get_input_nodes());
+}
+
+// The sub-graph construction algorithm is as follows
+// For each node, check its predecessors, if 
+// - all predecessors in sub-graphs belong to the same sub-graph (graph ID), then extend the sub-graph to include the current node.
+//   Predecessors outside sub-graphs are marked as input to the sub-graph. 
+// - predecessors in sub-graphs belong to different sub-graphs, then merge all the sub-graphs into one, and add current node to it.
+//   Predecessors outside sub-graphs are marked as input to the sub-graph. 
+//
+// For each sub-graph construction, build a CompiledKernel(CK) node around it as follows
+// - all inputs edges to the sub-graph are cloned as inputs to CK node as well.
+// - all outputs edges from the sub-graph are removed and added as outputs to CK node instead.
+// - CK will internally have lists record graph nodes, and graph output nodes.
 bool MLIRSubgraphExtractionPass::run_on_function(std::shared_ptr<Function> func)
 {
-    // Create a CompiledKernel for all the ops in the function, except Parameters and Results.
-    NodeVector ck_ops;
     for (auto op : func->get_ordered_ops())
     {
-        // All ops must be supported by MLIR compiler
+        NodeVector inputs;
+        int graph_id = -1;
+
+        // unsupported ops, skip
         if (!is_supported_mlir_op(op))
         {
-            return false;
+            continue;
+        }
+        if (TI(Parameter) == TI(*op) || TI(Result) == TI(*op))
```

Comment:
How do we plan to treat 'constants'? supported or unsupported?

Reply:
Right now constants are inputs to sub-graph since they are not supported?

Reply:
They are inputs yes, and shouldn't be supported, since there is no operation required for them. So MLIR code will just read from the input tensor.

Reply:
Thinking about it more, we should support constant ops in MLIR for perf reasons. We need to be able to do constant folding/prop, and have a constant tensor def op would enable that. Alternatively, we could leave it unsupported but those inputs as constant. I prefer the former.